### PR TITLE
sql,catalog/descs,gcjob,lease,importccl,backupccl: replace PublishMultiple

### LINKED ISSUE #50906: sql: PublishMultiple() and Publish() don't handle ErrDidntUpdateDescriptor correctly
`PublishMultiple()` and `Publish()` (on `lease.Storage`, soon to be `lease.Manager`) don't handle `ErrDidntUpdateDescriptor` correctly when it's returned from the descriptor update closure. The problem is that we return the error in the transaction closure before getting around to populating the in-memory map of updated descriptors:

https://github.com/cockroachdb/cockroach/blob/7458f24b1ceec9ca1bdfa34f2d6a198c750d39cd/pkg/sql/catalog/lease/lease.go#L424

So it's possible for these methods to return an empty map. This doesn't seem to pose any problems currently because we don't use the returned map of descriptors often, but it should be fixed.

### LINKED ISSUE #51417: sql,lease: `Publish`/`PublishMultiple` are bad APIs which can cause bugs
**Describe the problem**

The `lease.Manager` contains a method [`PublishMultiple`](https://godoc.org/github.com/cockroachdb/cockroach/pkg/sql/catalog/lease#Manager.PublishMultiple) which takes a set of IDs to modify and a closure to modify them and encapsulates creating an transaction and running retry loops. This method is utilized in schema changes (and in some cases in imports and restores).

One observation is that this method doesn't utilize anything interesting about the `Manager` object on which it resides. Worse yet is that this API takes a static set of IDs with no mechanism to update that set upon retry. 

In the case of schema changes, this set of IDs can refer to tables referenced as foreign keys, or, more recently, referenced types. A hazardous race which can occur is that the table becomes dropped while the schema change is operating. If this happens, there is no longer an invariant that the referenced tables exist. In practice, due to the fact that tables wait some time for GC, this hazard is unlikely to come up in reference to fk relationships. With types however there is no such waiting period.

What this all points out is that the PublishMultiple interface is bad. It probably should not live in the lease package, or at least, should not be a method, and it should not be responsible for creating a transaction.

**Describe the solution you'd like**
The main solution I'd like to see is that we either nix `Publish`/`PublishMultiple` entirely, or we have them take a transaction and a set of read descriptors from that transaction rather than a set of IDs.

### PR DESCRIPTION
**Pull request by ajwerner**
This series of commits removed `lease.Manager.PublishMultiple()` and its variants and replaces them with a new transactional wrapper around the `descs.Collection`.

Fixes #50906
Fixes #51417

This is WIP on the road to giving a similar treatment to restore and import as a part of #53378.

### CODE REVIEW

**Comment by ajwerner** [comment]
@lucy-zhang I would still like to add more unit testing to this but I think it is ready for review. I'll admit that it touches more than is maybe ideal but I think it leaves us in a happier place. LMKWYT

**Comment by ajwerner** [comment]
@pbardea it'd be wonderful if you were willing to take a look at the last commit in particular. It's in many ways the meat of this change.

**Comment by ajwerner** [comment]
Okay I've rebased on top of #53848

**Review by pbardea — COMMENTED** [review]
Did a first pass on the last commit and it looks good to me
Thanks for rebasing; sorry for not CC'ing you earlier.

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 5 of 5 files at r6, 2 of 2 files at r7, 1 of 5 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-:-MGK4i2e9lokQkwMCwpz:b-curooq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @lucy-zhang, and @pbardea)

___
*[pkg/ccl/backupccl/restore_job.go, line 810 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MGK8S9p3i7v-H1zoqoo:-MGK8S9p3i7v-H1zoqop:b-sftw98) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2c5dbf663ce58b4c119ed36e370a330fa95749/pkg/ccl/backupccl/restore_job.go#L810)):*
> ```Go
> 		case catalog.TableDescriptor:
> 			mut := tabledesc.NewCreatedMutable(*desc.TableDesc())
> 			mut.Version = 1
> ```

I think it's interesting that we need to reset the version of ingested descriptors. Is there a comment that explains why this is needed? Maybe a short note around here would be helpful.
___
*[pkg/ccl/backupccl/restore_job.go, line 1355 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MGK4cZL39HT6lbx9P5S:-MGK4cZMA0jfTRfFy4p2:btx2v11) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2c5dbf663ce58b4c119ed36e370a330fa95749/pkg/ccl/backupccl/restore_job.go#L1355)):*
> ```Go
> 		// it's not really clear what to do. Just log and carry on.
> 		if got, exp := mutableTables[i].Version, details.TableDescs[i].Version; got != exp {
> 			log.Errorf(ctx, "version changed for restored descriptor before "+
> ```

Let's also log the descriptor ID here?

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53727)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @lucy-zhang, and @pbardea)

___
*[pkg/ccl/backupccl/restore_job.go, line 810 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MGK8S9p3i7v-H1zoqoo:-MGKAoWg70d2RXyjGk7F:bwmhi28) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2c5dbf663ce58b4c119ed36e370a330fa95749/pkg/ccl/backupccl/restore_job.go#L810)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

I think it's interesting that we need to reset the version of ingested descriptors. Is there a comment that explains why this is needed? Maybe a short note around here would be helpful.
</blockquote></details>

I made it so we don't need to do this. I'll leave it to Lucy to do this in https://github.com/cockroachdb/cockroach/pull/53539.
___
*[pkg/ccl/backupccl/restore_job.go, line 1355 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MGK4cZL39HT6lbx9P5S:-MGKCLw88nfwjGEkB5jY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2c5dbf663ce58b4c119ed36e370a330fa95749/pkg/ccl/backupccl/restore_job.go#L1355)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Let's also log the descriptor ID here?
</blockquote></details>

Done.

File: pkg/ccl/importccl/import_stmt.go [hunk 1]
```diff
@@ -1319,15 +1316,16 @@ func (r *importResumer) publishTables(ctx context.Context, execCfg *sql.Executor
 	log.Event(ctx, "making tables live")
 
 	// Needed to trigger the schema change manager.
```
Comment (lucy-zhang) [inline pkg/ccl/importccl/import_stmt.go:1318]: nit: this comment is very outdated

File: pkg/ccl/backupccl/restore_job.go [hunk 2]
```diff
@@ -1350,21 +1335,28 @@ func (r *restoreResumer) dropDescriptors(
...
 	// Collect the tables into mutable versions.
 	mutableTables := make([]*tabledesc.Mutable, len(details.TableDescs))
 	for i := range details.TableDescs {
-		mutableTables[i] = tabledesc.NewExistingMutable(*details.TableDescs[i])
+		var err error
+		mutableTables[i], err = descriptors.GetMutableTableVersionByID(ctx, details.TableDescs[i].ID, txn)
+		if err != nil {
+			return err
+		}
+		// Ensure that the version matches we expect. In the case that it doesn't,
```
Comment (lucy-zhang) [inline pkg/ccl/backupccl/restore_job.go:1348]: nit: `what we expect`

Comment (lucy-zhang) [inline pkg/ccl/backupccl/restore_job.go:1348]: IIUC the descriptors could either still be offline or be public at this point. If they're already public, we can't actually make any guarantees about their versions, right?

File: pkg/sql/catalog/descs/txn.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+//
+// The passed transaction is pre-emptively anchored to the system config key on
+// the system tenant.
+func Txn(
+	ctx context.Context,
+	settings *cluster.Settings,
+	leaseMgr *lease.Manager,
+	ie sqlutil.InternalExecutor,
+	db *kv.DB,
+	f func(ctx context.Context, txn *kv.Txn, descriptors *Collection) error,
```
Comment (lucy-zhang) [inline pkg/sql/catalog/descs/txn.go:45]: nit: I might prefer this argument be `descCollection` or something. `descriptors` always makes me think we're calling functions in a `descriptors` package. I don't feel strongly about this at all.

**Review by thoszhang — COMMENTED** [review]
I think you squashed two commits into your last commit and didn't update the default message.

**Review by ajwerner — COMMENTED** [review]
Rebased and RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-:-MHB9Pwu0iBZj5Iob0G_:b-1ti1ar)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @lucy-zhang, and @pbardea)

___
*[pkg/ccl/backupccl/restore_job.go, line 1348 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MGJzq6jDufUWZ1L06Jd-r30-1348:-MHAfSSV8BU3IGXPnTTD:bspifj8) ([raw file](https://github.com/cockroachdb/cockroach/blob/16a5cda799a177750573dab46dd9864d084e1e05/pkg/ccl/backupccl/restore_job.go#L1348)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

IIUC the descriptors could either still be offline or be public at this point. If they're already public, we can't actually make any guarantees about their versions, right?
</blockquote></details>

Done. Only check now if we haven't published the descriptors.
___
*[pkg/sql/schema_changer.go, line 904 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MG7IuU9UdrgdzqPtcdt-r27-904:-MHAWOim9HEnQlEYPFSY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfdc19333c229f625f9d621c93c680514f9e6932/pkg/sql/schema_changer.go#L904)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

nit: update this comment
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descs/txn.go, line 45 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MG7IuU9UdrgdzqPtce0-r26-45:-MHAWlzY3E77xlkQp9yi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/374364eda1cb3dbb2707a397bbec9b111d731105/pkg/sql/catalog/descs/txn.go#L45)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

nit: I might prefer this argument be `descCollection` or something. `descriptors` always makes me think we're calling functions in a `descriptors` package. I don't feel strongly about this at all.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/lease/helpers_test.go, line 123 at r29](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MGJzq6jDufUWZ1L06Jb-r29-123:-MHAZh0s4XPPCzGEmVW1:b-dkyl7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8237ac6d18d26fe7af3794b33182e6cdd340ff4c/pkg/sql/catalog/lease/helpers_test.go#L123)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

We can get rid of this now, right? It doesn't even seem to be used in the tests besides a test for the `ErrDidntUpdateDescriptor` functionality itself.
</blockquote></details>

sure, done, I'd love to kill Publish and PublishMultiple too but that can be follow-up. At least they now live in testing.

**Review by thoszhang — APPROVED** [review]
> Reviewed 1 of 2 files at r21, 2 of 2 files at r22, 4 of 5 files at r25, 2 of 2 files at r26, 5 of 5 files at r27, 2 of 2 files at r28, 2 of 2 files at r29, 4 of 4 files at r30, 1 of 2 files at r32, 2 of 2 files at r33, 5 of 7 files at r36, 2 of 2 files at r37, 5 of 5 files at r38, 2 of 2 files at r39, 3 of 6 files at r40, 5 of 5 files at r41, 2 of 2 files at r42.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53727)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @pbardea)

___
*[pkg/ccl/backupccl/restore_job.go, line 999 at r42](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MHBko7a3Gn3H7H7zuXC:-MHBko7a3Gn3H7H7zuXD:br25gss) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9916c50a4f57e4df413ba9d0ddf697500dd65c/pkg/ccl/backupccl/restore_job.go#L999)):*
> ```Go
> 						}
> 						typDesc.AddReferencingDescriptorID(table.GetID())
> 						// TODO (lucy): I think we should be incrementing the version for the
> ```

We can get rid of this comment now.
___
*[pkg/ccl/backupccl/restore_job.go, line 1262 at r42](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MHC71QO9vpJ-ApHAOCO:-MHC71QP9K5rvsmNZ7bN:ba62m03) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9916c50a4f57e4df413ba9d0ddf697500dd65c/pkg/ccl/backupccl/restore_job.go#L1262)):*
> ```Go
> func (r *restoreResumer) publishDescriptors(
> 	ctx context.Context, txn *kv.Txn, descsCol *descs.Collection,
> ) (newDescriptorChangeJobs []*jobs.StartableJob, err error) {
> ```

Returning the jobs is kind of weird because the jobs just wait for leases to drain, but we don't actually wait for the jobs to finish, right? I'm fine with this as-is since I think it's good consistency-wise for schema changes to always be tied to jobs, but the extra complexity for no good reason is kind of annoying.
___
*[pkg/ccl/backupccl/restore_job.go, line 1387 at r42](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MHBmDbJ0NQBz5S_AAlo:-MHBmDbJ0NQBz5S_AAlp:b-fqsjnr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9916c50a4f57e4df413ba9d0ddf697500dd65c/pkg/ccl/backupccl/restore_job.go#L1387)):*
> ```Go
> 
> 	if err := txn.Run(ctx, b); err != nil {
> 		return newDescriptorChangeJobs, errors.Wrap(err, "publishing tables")
> ```

Is this intentional?
___
*[pkg/sql/catalog/descs/txn.go, line 45 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MG7IuU9UdrgdzqPtce0-r26-45:-MHBpE8i-k0XVxa3gjG2:bbs25uq) ([raw file](https://github.com/cockroachdb/cockroach/blob/374364eda1cb3dbb2707a397bbec9b111d731105/pkg/sql/catalog/descs/txn.go#L45)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Could you also update this in all the other files?

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53727)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @lucy-zhang, and @pbardea)

___
*[pkg/ccl/backupccl/restore_job.go, line 999 at r42](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MHBko7a3Gn3H7H7zuXC:-MHCVlU4DdJEE20BO3hI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9916c50a4f57e4df413ba9d0ddf697500dd65c/pkg/ccl/backupccl/restore_job.go#L999)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

We can get rid of this comment now.
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/restore_job.go, line 1262 at r42](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MHC71QO9vpJ-ApHAOCO:-MHCWF6i1btUw_Kb2UFL:b-c8nnga) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9916c50a4f57e4df413ba9d0ddf697500dd65c/pkg/ccl/backupccl/restore_job.go#L1262)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Returning the jobs is kind of weird because the jobs just wait for leases to drain, but we don't actually wait for the jobs to finish, right? I'm fine with this as-is since I think it's good consistency-wise for schema changes to always be tied to jobs, but the extra complexity for no good reason is kind of annoying.
</blockquote></details>

it's important because you can't start them until the transaction commits and the transaction is handled above this call. Added a comment. 
___
*[pkg/ccl/backupccl/restore_job.go, line 1387 at r42](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MHBmDbJ0NQBz5S_AAlo:-MHCWAUF1Y0JHmWKzila:biuwarn) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9916c50a4f57e4df413ba9d0ddf697500dd65c/pkg/ccl/backupccl/restore_job.go#L1387)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Is this intentional?
</blockquote></details>

this being? returning the jobs, yes, there's a defer to clean them up. 
___
*[pkg/sql/catalog/descs/txn.go, line 45 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MG7IuU9UdrgdzqPtce0-r26-45:-MHCWNt50WrhNfla7Rcu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/374364eda1cb3dbb2707a397bbec9b111d731105/pkg/sql/catalog/descs/txn.go#L45)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Could you also update this in all the other files?
</blockquote></details>

Done.

**Review by thoszhang — APPROVED** [review]
> Reviewed 1 of 16 files at r43.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53727)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @lucy-zhang, and @pbardea)

___
*[pkg/ccl/backupccl/restore_job.go, line 1262 at r42](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MHC71QO9vpJ-ApHAOCO:-MHCZ7WeD_pmFmL53eRJ:b-lewgce) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9916c50a4f57e4df413ba9d0ddf697500dd65c/pkg/ccl/backupccl/restore_job.go#L1262)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

it's important because you can't start them until the transaction commits and the transaction is handled above this call. Added a comment. 
</blockquote></details>

Sorry, my point wasn't about returning the jobs per se, it was more that we even start the jobs at all when the jobs do nothing except wait for leases to drain, but we don't wait for the jobs to finish within this job. I'm fine with this as-is, but it's just unsatisfying.
___
*[pkg/ccl/backupccl/restore_job.go, line 1387 at r42](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MHBmDbJ0NQBz5S_AAlo:-MHCZdXC7owtaa5P_Vqs:b2xiksc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9916c50a4f57e4df413ba9d0ddf697500dd65c/pkg/ccl/backupccl/restore_job.go#L1387)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this being? returning the jobs, yes, there's a defer to clean them up. 
</blockquote></details>

Sorry, I was looking at an earlier revision, never mind.

**Comment by ajwerner** [comment]
TFTR!

bors r+

**Pull request merged**